### PR TITLE
tuxclocker: init at 1.4.0

### LINKED ISSUE #262668: Package request: tuxclocker
**Project description**

TuxClocker is a hardware controlling and monitoring program. TuxClocker consists of a DBus daemon and a Qt GUI that uses the daemon.

It allows you to e.g. set a custom fan curve and overclock your CPU and GPU.

**Metadata**

* homepage URL: https://github.com/Lurkki14/tuxclocker
* source URL: https://github.com/Lurkki14/tuxclocker/tree/master/src
* license: gpl3
* platforms: linux

### PR DESCRIPTION
**Pull request by Lurkki14**
## Description of changes
Added https://github.com/Lurkki14/tuxclocker and a NixOS module

Fixes #262668

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Lurkki14** [comment]
@ofborg build tuxclocker-without-unfree

File: nixos/modules/services/misc/tuxclocker.nix [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+with lib;
+
+let
+  cfg = config.programs.tuxclocker;
+in
+{
+  options.programs.tuxclocker = {
+    enable = mkEnableOption (lib.mdDoc ''
+      TuxClocker, a hardware control and monitoring program.
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/tuxclocker.nix:11]:
```suggestion
      TuxClocker, a hardware control and monitoring program
```
that is already added by mkEnableOption

File: nixos/modules/services/misc/tuxclocker.nix [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    '');
+
+    enableAMD = mkEnableOption (lib.mdDoc ''
+      Enable AMD GPU controls.
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/tuxclocker.nix:15]:
```suggestion
      AMD GPU controls.
```

File: nixos/modules/services/misc/tuxclocker.nix [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      Sets the `amdgpu.ppfeaturemask` kernel parameter to 0xfffd7fff to enable all TuxClocker controls.
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/tuxclocker.nix:16]:
```suggestion
      Sets the `amdgpu.ppfeaturemask` kernel parameter to 0xfffd7fff to enable all TuxClocker controls
```

File: nixos/modules/services/misc/tuxclocker.nix [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    '');
+
+    enabledNVIDIADevices = mkOption {
+      type = (types.listOf types.int);
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/tuxclocker.nix:20]:
```suggestion
      type = types.listOf types.int;
```

File: nixos/modules/services/misc/tuxclocker.nix [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+      default = [ ];
+      example = [ 0 1 ];
+      description = lib.mdDoc (''
+        Enable NVIDIA GPU controls for a device by index.
+        Sets the `Coolbits` Xorg option to enable all TuxClocker controls.
+      '');
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/tuxclocker.nix:26]:
```suggestion
      description = lib.mdDoc ''
        Enable NVIDIA GPU controls for a device by index.
        Sets the `Coolbits` Xorg option to enable all TuxClocker controls.
      '';
```

File: nixos/modules/services/misc/tuxclocker.nix [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    };
+
+    useUnfree = mkOption {
+      type = types.bool;
+      default = true;
+      example = false;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/tuxclocker.nix:32]:
```suggestion
      default = false;
      example = true;
```
free by default is a good default for nixos

File: nixos/modules/services/misc/tuxclocker.nix [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      description = lib.mdDoc (''
+        Whether to use components requiring unfree dependencies.
+        Disabling this allows you to get everything from the binary cache.
+      '');
+    };
+  };
+
+  config = mkIf cfg.enable (lib.mkMerge [
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/tuxclocker.nix:40]: I don't think we need a mkMerge here.

File: nixos/modules/services/misc/tuxclocker.nix [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    {
+      environment.systemPackages = [
+        (if cfg.useUnfree then pkgs.tuxclocker else pkgs.tuxclocker-without-unfree)
+      ];
+
+      services.dbus.packages = [
+        (if cfg.useUnfree then pkgs.tuxclocker else pkgs.tuxclocker-without-unfree)
+      ];
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/tuxclocker.nix:48]: We should put the package into a let in to not duplicate the condition

File: pkgs/applications/misc/tuxclocker/default.nix [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "tuxclocker";
+  version = "1.4.0";
+
+  src = fetchFromGitHub {
+    fetchSubmodules = true;
+    owner = "Lurkki14";
+    repo = "tuxclocker";
+    rev = "${finalAttrs.version}";
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/tuxclocker/default.nix:25]:
```suggestion
    owner = "Lurkki14";
    repo = "tuxclocker";
    fetchSubmodules = true;
    rev = "${finalAttrs.version}";
```
nit

File: pkgs/applications/misc/tuxclocker/default.nix [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    hash = "sha256-8dtuZXBWftXNQpqYgNQOayPGfvEIu9QfbqDShfkt1qA=";
+  };
+
+  # Meson doesn't find boost without these
+  BOOST_INCLUDEDIR = "${lib.getDev boost}/include";
+  BOOST_LIBRARYDIR = "${lib.getLib boost}/lib";
+
+  nativeBuildInputs = [
+    git
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/tuxclocker/default.nix:34]: Why do we need git?

File: pkgs/applications/misc/tuxclocker/default.nix [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    boost
+    qtbase
+    qtcharts
+  ];
+
+  postInstall = ''
+    wrapProgram "$out/bin/tuxclockerd" \
+      --prefix "TEXTDOMAINDIR" : "${tuxclocker-plugins}/share/locale" \
+      --prefix "TUXCLOCKER_PLUGIN_PATH" : "${tuxclocker-plugins}/lib/tuxclocker/plugins" \
+      --prefix "PYTHONPATH" : "$(toPythonPath ${python3.pkgs.hwdata})"
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/tuxclocker/default.nix:53]:
```suggestion
      --prefix "PYTHONPATH" : "${python3.pkgs.hwdata}/${python3.sitePackages}"
```
we can avoid the subshell here

File: pkgs/by-name/tu/tuxclocker-nvidia-plugin/package.nix [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+, boost
+, libX11
+, libXext
+, linuxPackages
+, openssl
+, tuxclocker-plugins
+}:
+
+stdenv.mkDerivation {
+  name = "tuxclocker-nvidia-plugin";
```
Comment (SuperSandro2000) [inline pkgs/by-name/tu/tuxclocker-nvidia-plugin/package.nix:12]:
```suggestion
  pname = "tuxclocker-nvidia-plugin";
```

File: pkgs/by-name/tu/tuxclocker-plugins/package.nix [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+}:
+
+stdenv.mkDerivation {
+  inherit (tuxclocker) src version meta BOOST_INCLUDEDIR BOOST_LIBRARYDIR;
+
+  pname = "tuxclocker-plugins";
+
+  nativeBuildInputs = [
+    gettext
+    git
```
Comment (SuperSandro2000) [inline pkgs/by-name/tu/tuxclocker-plugins/package.nix:23]: Why do we need git here?

File: pkgs/applications/misc/tuxclocker/default.nix (hunk 10 above)
Comment (Lurkki14) [reply pkgs/applications/misc/tuxclocker/default.nix:34]: It's used by meson to determine version string

File: pkgs/by-name/tu/tuxclocker-plugins/package.nix (hunk 13 above)
Comment (Lurkki14) [reply pkgs/by-name/tu/tuxclocker-plugins/package.nix:23]: Same here, though the result is actually not needed when building the plugins. Alternative would be to patch it

File: nixos/modules/services/misc/tuxclocker.nix [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+              Option "Coolbits" "31"
+              Identifier "Device-nvidia[${toString i}]"
+            EndSection
+          '');
+        in
+          concatStrings (map configSection cfg.enabledNVIDIADevices);
+
+        # https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/gpu/drm/amd/include/amd_shared.h#n207
+        # Enable everything modifiable in TuxClocker
+        boot.kernelParams = mkIf cfg.enableAMD [ "0xfffd7fff" ];
```
Comment (delroth) [inline nixos/modules/services/misc/tuxclocker.nix:69]: Shouldn't this be amdgpu.ppfeaturemask=...?

Comment (Lurkki14) [reply nixos/modules/services/misc/tuxclocker.nix]: Indeed, good catch

**Review by delroth — APPROVED** [review]

File: nixos/doc/manual/release-notes/rl-2405.section.md [hunk 15]
```diff
@@ -23,6 +23,8 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 
 - [Clevis](https://github.com/latchset/clevis), a pluggable framework for automated decryption, used to unlock encrypted devices in initrd. Available as [boot.initrd.clevis.enable](#opt-boot.initrd.clevis.enable).
 
+- [TuxClocker](https://github.com/Lurkki14/tuxclocker), a hardware control and monitoring program. Available as [programs.tuxclocker](#programs.tuxclocker.enable).
```
Comment (delroth) [inline nixos/doc/manual/release-notes/rl-2405.section.md:26]: This link to the option is broken, according to the failing manual CI check. I think it needs to be `#opt-programs.tuxclocker.enable`?

**Review by delroth — CHANGES_REQUESTED** [review]
Oops, missed that CI failed while I was re-checking this.

**Review by delroth — APPROVED** [review]

**Comment by Lurkki14** [comment]
@ofborg build tuxclocker-without-unfree

**Pull request merged**